ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStoreV2.java

```diff
@@ -63,10 +79,145 @@ public CloudBlobStoreV2(VerifiableProperties properties, MetricRegistry metricRe
+    if (messageInfos.size() < 2) {
+      return;
+    }
+    HashMap<String, MessageInfo> seenKeys = new HashMap<>();
+    for (MessageInfo messageInfo : messageInfos) {
+      // Check for duplicates. We cannot continue as stream could be corrupted.
+      // Server always sends one message per key that captures full state of the key including ttl-update & delete.
+      // Emit a log and metric for investigation as it indicates a flaw in (de)serialization or replication.
+      // FIXME: This check should be in replication layer to avoid multiple implementations at Store level.
+      String blobId = messageInfo.getStoreKey().getID();
+      if (seenKeys.containsKey(blobId)) {
+        throw new IllegalArgumentException(
+            String.format("Duplicate message in replication-batch, original message = %s, duplicate message = %s",
+                seenKeys.get(blobId).toString(), messageInfo));
+      }
+      seenKeys.put(blobId, messageInfo);
+    }
+  }
+
+  /**
+   * Uploads Ambry blob to Azure blob storage.
+   * If blob already exists, then it catches BLOB_ALREADY_EXISTS exception and proceeds to next blob.
+   * It fails for any other exception.
+   *
+   * A blob could be a simple blob, in which case its blob-id points to one data chunk.
+   * A blob could be a composite blob, in which case its blob-id points to a metadata chunk which in turn
+   * points to several data chunks. A metadata blob is usually a few bytes long.
+   *
+   * Consider a composite blob with 64 MB of data and blob-id Tk2ZGYtNDk2Mi1hZ. Suppose its metadata blob is 100 bytes.
+   * A get-blob-info(Tk2ZGYtNDk2Mi1hZ) request to Ambry will display blob size as 64 MB, and not 100 bytes.
+   * But the blob in Azure blob storage corresponding to Tk2ZGYtNDk2Mi1hZ will be 100 bytes because it is a
+   * metadata chunk. There will be 16 data chunks in Azure storage each of 4 MB that belong to this composite blob.
+   *
+   * @param messageSetToWrite The message set to write to the store
+   *                          Only the StoreKey, OperationTime, ExpirationTime, LifeVersion should be used in this method.
+   * @throws StoreException
+   */
   @Override
   public void put(MessageWriteSet messageSetToWrite) throws StoreException {
-    // TODO
-  }
+    MessageSievingInputStream messageSievingInputStream =
+        (MessageSievingInputStream) ((MessageFormatWriteSet) messageSetToWrite).getStreamToWrite();
+
+    // This is a list of blob-ids from remote server
+    List<MessageInfo> messageInfos = messageSievingInputStream.getValidMessageInfoList();
+    checkDuplicates(messageInfos);
+
+    // This is a list of data-streams associated with blob-ids. i-th stream is for i-th blob-id.
+    // This allows us to pass stream to azure-sdk and let it handle read() logic.
+    List<InputStream> messageStreamList = messageSievingInputStream.getValidMessageStreamList();
+    if (messageInfos.size() != messageStreamList.size()) {
```

Comment:
The separation of msg infos and msg stream into two lists happens only in `handleGetResponse` L1232 and immediately after that is `applyPut`. There are no checks in between that assert the length of the two lists are the same. So whether this check is `handleGetResponse` or `applyPut`, it happens just once. However, unequal lengths would be really really rare because the addition to each list happens together like this
```
      sievedMessageInfoList.add(inMsg.getMessageInfo());
      msgStreamList.add(inMsg.getStream());
```